src/keyDerivation.c

```diff
@@ -1,72 +1,310 @@
+    // Crypto assets to be cleared
     private_key_t privateKey;
-    BEGIN_TRY {
-        TRY {
-            derivePrivateKey(pathSpec, &privateKey);
-            // We should do cx_ecfp_generate_pair here, but it does not work in SDK < 1.5.4,
-            // should work with the new SDK
-            io_seproxyhal_io_heartbeat();
-            CX_THROW(cx_ecfp_init_public_key_no_throw(CX_CURVE_SECP256K1, NULL, 0, publicKey));
-            CX_THROW(cx_ecfp_generate_pair_no_throw(CX_CURVE_SECP256K1,
-                                                    publicKey,
-                                                    &privateKey,
-                                                    1));  // 1 - private key preserved
-            io_seproxyhal_io_heartbeat();
+    uint8_t V[33];
+    uint8_t K[32];
+    explicit_bzero(&privateKey, SIZEOF(privateKey));
+    explicit_bzero(V, SIZEOF(V));
+    explicit_bzero(K, SIZEOF(K));
+    explicit_bzero(signature, signatureLen);
+
+    // Variable for CRYPTO_ error handling macros
+    uint16_t error_to_return = ERR_ASSERT;
+
+    CRYPTO_ASSERT(signatureLen >= 200);
+
+    CRYPTO_FORWARD_ERROR(derivePrivateKey(witnessPath, &privateKey));
+    TRACE("privateKey.d:");
+    TRACE_BUFFER(privateKey.d, privateKey.d_len);
+
+    // We sign the hash
+    int tries = 0;
+    // Loop until a candidate matching the canonical signature is found
+    for (;;) {
+        if (tries == 0) {
+            CRYPTO_FORWARD_ERROR(rng_rfc6979(signature + 100,
+                                             hashBuf,
+                                             privateKey.d,
+                                             privateKey.d_len,
+                                             SECP256K1_N,
+                                             32,
+                                             V,
+                                             K));
+        } else {
+            CRYPTO_FORWARD_ERROR(
+                rng_rfc6979(signature + 100, hashBuf, NULL, 0, SECP256K1_N, 32, V, K));
         }
-        FINALLY {
-            explicit_bzero(&privateKey, SIZEOF(privateKey));
+        uint32_t infos;
+
+        size_t sig_len_ = 100;
+        CRYPTO_CX_CHECK(cx_ecdsa_sign_no_throw(&privateKey,
+                                               CX_NO_CANONICAL | CX_RND_PROVIDED | CX_LAST,
+                                               CX_SHA256,
+                                               hashBuf,
+                                               32,
+                                               signature + 100,
+                                               &sig_len_,
+                                               &infos));
+        TRACE_BUFFER(signature + 100, 100);
+
+        if ((infos & CX_ECCINFO_PARITY_ODD) != 0) {
+            signature[100] |= 0x01;
+        }
+        signature[0] = 27 + 4 + (signature[100] & 0x01);
+        ecdsa_der_to_sig(signature + 100, signature + 1);
```

Comment:
Should use `cx_ecdsa_sign_rs_no_throw` to avoid having to handle the conversion from der to sign in the app code.

Reply:
The signature produced by cx_ecdsa_sign_no_throw is still needed for rng_rfc6979 function for the next try in case check_canonical test fails.

Reply:
Indeed I missed it sorry.